schema.md

```diff
@@ -0,0 +1,136 @@
+# Sparkle Motion schema (WIP)
+
+All record types have `createdAt` (and maybe `modifiedAt`?).
+
+### User
+
+```js
+{
+  "id": ___,
+  "username": "xian",
+  "password": "...",      // salted hash of pw
+  "name": "Full Name",
+  "priority": 123.4,      // e.g. for locking out other users during a performance
+}
+```
+
+### Gadget Presets
+
+Gadgets offer the ability to save their current values as a user preset, or to load presets.
+
+```js
+{
+  "id": ___,
+  "userId": ___,
+  "name": "Pride Flag",
+  "type": "Palette",      // the type of the gadget
+  "data": { ... },        // type-dependent
+}
+```
+
+#### Color Gadget Data
+```js
+{
+  "color": 0xfedbca,
+}
+```
+
+#### Palette Gadget Data
+```js
+{
+  "type": "analogous" | "monochromatic" | "triad" | "complementary" | "compound" | "shades" | "custom",
+
+  // e.g. for type "custom":
+  "values": [0xE70000, 0xFF8C00, 0xFFEF00, 0x00811F, 0x0044FF, 0x760089],
+}
+
+```
+
+#### Slider Gadget Data
+```js
+{
+  "value": 0.7325,
+}
+```
+
+### Show Presets
+
+All of the current gadget settings for a show can be saved as a preset, which then shows up in the show selector.
+
+```js
+{
+  "userId": ___,
+  "name": "Awesome",
+  "showType": "TexturesAndStuff",
+  "gadgets": [
+    {
+      "id": "texture",    // the show's id for the gadget
+      "type": "Texture",  // the type of the gadget
+      "data": { "file": "/textures/tiger-print.png" }, // type-dependent
+    }
+  ]
+}
+```
+
+### Moving Head Positions
+
+These represent a focal point for the moving heads.
+
+Currently I'm thinking the eyes work like this: in the UI, you can specify a base position for moving heads. When
+`show.nextFrame()` is called, each moving head is already set to that base position, and it can then be altered
+arbitrarily by the show.
+
+These are saved position presets. 
+
+```js
+{
+  "name": "Headlight Mode",
+  "target": { "x": ___, "y": ___, "z": ___ },
+}
+
+```
+
+These could also be represented using `pan`/`tilt`, but then each head would need its own entry. That might make sense
+for disco ball mode, where only the left eye can hit the disco balls (and the other might point straight up).
+
+### Panel Mapping Data
+
+These are collected on each panel mapping run.
+
+Maybe needs more consideration:
+* how does Pinky determine which one is authoritative? (I think we want the one with the greatest area.)
+* how do we do admin overrides (e.g. a Brain dies and we need to swap a new one in)?
```

Comment:
So, then, does the concept of `outdatedAt` come into play when a brain dies or fails or otherwise needs replaced? In such an event, then only changing parameter for this type would be `brainId`, because the pixels would likely remain affixed identically to the panel? Or, maybe not, depending on how the boards are attached to the LED run(s).

Reply:
Yeah, exactly what I'm thinking. I'm not generally a fan of mutating system-generated data, but I guess that's probably the easiest thing to do here.

Reply:
Yeah, exactly what I'm thinking. I'm not generally a fan of mutating system-generated data, but I guess that's probably the easiest thing to do here.